src/main/java/com/direwolf20/buildinggadgets/common/entities/BlockBuildEntity.java

```diff
@@ -133,23 +127,31 @@ protected void writeAdditional(NBTTagCompound compound) {
 
     @Override
     protected void onSetDespawning() {
-        if (targetPos != null && setBlock != null && (getToolMode() == 1)) {
-            if (getUsingConstructionPaste()) {
-                world.setBlockState(targetPos, BGBlocks.constructionBlock.getDefaultState());
-                TileEntity te = world.getTileEntity(targetPos);
-                if (te instanceof ConstructionBlockTileEntity) {
-                    ((ConstructionBlockTileEntity) te).setBlockState(setBlock, actualSetBlock);
+        if (world.isRemote || targetPos == null || setBlock == null)
+            return;
+
+        switch (getToolMode()) {
+            case MODE_PLACE:
+                if (isUsingPaste()) {
+                    world.setBlockState(targetPos, BGBlocks.constructionBlock.getDefaultState());
+                    TileEntity te = world.getTileEntity(targetPos);
+                    if (te instanceof ConstructionBlockTileEntity) {
+                        ((ConstructionBlockTileEntity) te).setBlockState(setBlock, setBlock);
+                    }
+                    world.spawnEntity(new ConstructionBlockEntity(world, targetPos, false));
+                } else {
+                    world.setBlockState(targetPos, setBlock);
+                    BlockPos upPos = targetPos.up();
+                    world.getBlockState(targetPos).neighborChanged(world, targetPos, world.getBlockState(upPos).getBlock(), upPos);
                 }
-                world.spawnEntity(new ConstructionBlockEntity(world, targetPos, false));
-            } else {
-                world.setBlockState(targetPos, setBlock);
-                BlockPos upPos = targetPos.up();
-                world.getBlockState(targetPos).neighborChanged(world, targetPos, world.getBlockState(upPos).getBlock(), upPos);
-            }
-        } else if (targetPos != null && setBlock != null && getToolMode() == 2) {
-            world.setBlockState(targetPos, Blocks.AIR.getDefaultState());
-        } else if (targetPos != null && setBlock != null && getToolMode() == 3) {
-            world.spawnEntity(new BlockBuildEntity(world, targetPos, spawnedBy, originalSetBlock, 1, actualSetBlock, getUsingConstructionPaste()));
+                break;
+            case MODE_REMOVE:
+                world.setBlockState(targetPos, Blocks.AIR.getDefaultState());
+                break;
+            case MODE_REPLACE:
+                world.spawnEntity(new BlockBuildEntity(world, targetPos, spawnedBy, originalSetBlock, MODE_PLACE, isUsingPaste()));
+                break;
```

Comment:
The enum would make this switch unnecessary, if it just added this as an abstract private Method, implemented apropiatly for the different cases...